### PR TITLE
chore: Catch ruamel.yaml Duplicate Key Errors and format for users [DET-5542]

### PR DESCRIPTION
**Pull request by RehanSD**
## Description
When a duplicate key is found in the experiment config file, ruamel.yaml's exception is reported directly to the user, which looks like this:
```python
Traceback (most recent call last):
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/determined/cli/cli.py", line 275, in main
    parsed_args.func(parsed_args)
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/determined/cli/experiment.py", line 185, in create
    submit_experiment(args)
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/determined/common/api/authentication.py", line 27, in f
    return func(namespace)
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/determined/cli/experiment.py", line 125, in submit_experiment
    experiment_config = _parse_config_file_or_exit(args.config_file)
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/determined/cli/experiment.py", line 115, in _parse_config_file_or_exit
    experiment_config = yaml.safe_load(config_file.read())
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/ruamel/yaml/main.py", line 1105, in safe_load
    return load(stream, SafeLoader, version)
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/ruamel/yaml/main.py", line 1058, in load
    return loader._constructor.get_single_data()
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/ruamel/yaml/constructor.py", line 111, in get_single_data
    return self.construct_document(node)
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/ruamel/yaml/constructor.py", line 121, in construct_document
    for _dummy in generator:
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/ruamel/yaml/constructor.py", line 714, in construct_yaml_map
    value = self.construct_mapping(node)
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/ruamel/yaml/constructor.py", line 435, in construct_mapping
    return BaseConstructor.construct_mapping(self, node, deep=deep)
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/ruamel/yaml/constructor.py", line 253, in construct_mapping
    if self.check_mapping_key(node, key_node, mapping, key, value):
  File "/Users/rehandurrani/anaconda3/envs/det-test/lib/python3.6/site-packages/ruamel/yaml/constructor.py", line 284, in check_mapping_key
    raise DuplicateKeyError(*args)
ruamel.yaml.constructor.DuplicateKeyError: while constructing a mapping
  in "<unicode string>", line 1, column 1:
    description: constrained_adaptiv ...
    ^ (line: 1)
found duplicate key "checkpoint_storage" with value "{}" (original value: "{}")
  in "<unicode string>", line 7, column 1:
    checkpoint_storage:
    ^ (line: 7)

To suppress this check see:
    http://yaml.readthedocs.io/en/latest/api.html#duplicate-keys

Duplicate keys will become an error in future releases, and are errors
by default when using the new API.

Failed to create experiment
```
Instead of this, which exposes details of our internal codebase to users, this PR will catch the exception, and output the following:
```python
Error: failed while loading experiment config file examples/hp_search_benchmarks/darts_cifar10_pytorch/constrained_adaptive.yaml.
Duplicate Key Error: found duplicate key "learning_rate" with value "0.022" (original value: "0.025")
  in examples/hp_search_benchmarks/darts_cifar10_pytorch/constrained_adaptive.yaml, line 23, column 3:
      learning_rate: 0.022
      ^ (line: 23)
```

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

**Comment by shiyuann** [comment]
I think generally we should avoid depending our code on other libraries' error messages. When we upgrade libraries, this might break. The code is also hard to understand. Is there a better way to know which fields are duplicate from the error itself?

**Comment by RehanSD** [comment]
> I think generally we should avoid depending our code on other libraries' error messages. When we upgrade libraries, this might break. The code is also hard to understand. Is there a better way to know which fields are duplicate from the error itself?

Unfortunately they don't really give us any fields we can use. I originally wanted to try and do something more general, but that includes things like the link to the yaml docs etc, and @rb-determined-ai said for the sake of formatting, we should try and remove that as well

File: harness/determined/cli/experiment.py [hunk 1]
```diff
@@ -111,8 +111,67 @@ def read_git_metadata(model_def_path: pathlib.Path) -> Tuple[str, str, str, str]
...
+
+        To suppress this check see:
+            http://yaml.readthedocs.io/en/latest/api.html#duplicate-keys
+
+        Duplicate keys will become an error in future releases, and are errors
+        by default when using the new API.
+
+        Failed to create experiment
+    Output:
+        Error: failed while loading experiment config file constrained_adaptive.yaml.
```
Comment (RehanSD) [inline harness/determined/cli/experiment.py:141]: Should this instead say
```
Error: invalid experiment config file 
```?

Comment (RehanSD) [reply harness/determined/cli/experiment.py:141]: @shiyuann

**Review by shiyuann — APPROVED** [review]

File: harness/determined/cli/experiment.py (hunk 1 above)
Comment (shiyuann) [reply harness/determined/cli/experiment.py:141]: yeah this sounds more consistent

File: harness/determined/cli/experiment.py [hunk 2]
```diff
@@ -111,8 +111,67 @@ def read_git_metadata(model_def_path: pathlib.Path) -> Tuple[str, str, str, str]
...
+        line 23, column 3:
+            learning_rate: 0.022
+            ^ (line: 23)
+    ---------------------------------------------------------------------------------------------
+    """
+    if isinstance(e, yaml.constructor.DuplicateKeyError):
+        err_msg = (
+            "found"
+            + str(e)
+            .split("found")[1]  # Isolate part of error msg that says where the duplicate key is
```
Comment (rb-determined-ai) [inline harness/determined/cli/experiment.py:154]: Whoa, sorry.  I did suggest printing a better message, but I don't think parsing another library's output is an appropriate strategy.  Honestly better to just leave it, or pre-detect the issue.

**Review by shiyuann — APPROVED** [review]

File: harness/determined/util.py [hunk 3]
```diff
@@ -225,3 +227,49 @@ def wrapper(*arg: List, **kwarg: Dict) -> Any:
         return cast(T, wrapper)
 
     return make_wrapper
+
+
+def safe_load_yaml_with_exceptions(yaml_file: Union[io.FileIO, IO[Any]]) -> Any:
```
Comment (rb-determined-ai) [inline harness/determined/util.py:232]: Can you put this in `common.util`?

Just to make the dependencies a little less circular.  It used to be that `determined_common` was a totally separate package that wasn't allowed to depend on `determined` proper at all.  That's not strictly the case but I think this should be in `common.util` anyway.

**Review by rb-determined-ai — APPROVED** [review]
looks good, just move it to `common.util`.

File: harness/determined/util.py (hunk 3 above)
Comment (RehanSD) [reply harness/determined/util.py:232]: Sure - should I merge after that, or ask Shiyuan for a review again?

**Comment by RehanSD** [comment]
List of places where we keep `ruamel.yaml.safe_load` (for future reference):

* e2e_tests/tests/test_users.py::test_non_root_experiment
* e2e_tests/tests/template/test_template.py::test_set_template
* harness/determined/cli/command.py::parse_config (uses yaml.safe_load to cast values)
* harness/determined/cli/render.py::format_base64_as_yaml
* harness/determined/cli/template.py::_parse_config
* harness/tests/common/test_schemas.py::all_cases
* harness/tests/common/test_schemas.py::test_schemas
* tools/scripts/bumpenvs.py::main
* tools/scripts/refresh-ubuntu-amis.py::main
* tools/scripts/update-bumpenvs-yaml.py::main

**Review by shiyuann — APPROVED** [review]

**Pull request merged**